### PR TITLE
`General`: Notify users more reliably when there is a version update

### PR DESCRIPTION
**Pull request by bassner**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I translated all newly inserted strings into English and German.

### Description

This PR was intially started by @krusche after he noticed that the service worker doesn't reliably update the client without a bunch of reloads. This PR adds an interval check to the service worker to look for updates, and adds a new button to the outdated alert that, if clicked, tells the service worker to activate the new version.

### Steps for Testing

Prerequisites:
- Two commits with a build; the 2nd commit should have a different version that the first one in both package.json and gradle.
You can use these prepared releases:
https://bamboo.ase.in.tum.de/deploy/viewDeploymentVersionCommits.action?versionId=486310120 ( b26f4c0 )
https://bamboo.ase.in.tum.de/deploy/viewDeploymentVersion.action?versionId=486310121 ( e0ba3cc )

Steps:
- Deploy the first commit to a test server. Wait for it to launch completely, then load the page.
- Make sure the service worker is installed (check dev tools)
- Verify that the correct version is displayed in the page top bar
- Keep the page open. Do not reload.
- Deploy the second commit to the server. Wait. Do not reload.
- The page will probably show an alert about it being disconnected. That's normal. Do not reload.
- Wait for the blue "your app is outdated" alert to appear. Do not reload, don't click the button yet.
- Navigate around in the app. Verify that the blue outdated alert keeps being displayed.
- Close the alert by clicking the small "X".
- Continue to navigate around. Verify that the alert reappears after some time.
- Finally, press the "Update" button in the alert.
- Check that the app updated to the new version (check version in the top bar).

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### CODE REVIEW

File: src/main/webapp/i18n/de/alert.json [hunk 1]
```diff
@@ -5,6 +5,7 @@
             "disconnected": "Du bist nicht mehr mit dem Artemis Server verbunden. Bitte überprüfe deine Internetverbindung.",
             "reconnected": "Du bist wieder mit dem Artemis Server verbunden."
         },
-        "outdatedAlert": "<b>Du verwendest eine veraltete Version von Artemis:</b> Bitte lade die Artemis-Webseite neu, um Inkompatibilitäten zu vermeiden."
+        "outdatedAlert": "<b>Du verwendest eine veraltete Version von Artemis.</b> Bitte update die App auf die aktuellste Version:",
```
Comment (krusche) [inline src/main/webapp/i18n/de/alert.json:8]:
```suggestion
        "outdatedAlert": "<b>Du verwendest eine veraltete Version von Artemis.</b> Bitte aktualisiere Artemis auf die neueste Version.",
```

File: src/main/webapp/i18n/en/alert.json [hunk 2]
```diff
@@ -5,6 +5,7 @@
             "disconnected": "You are not connected to the Artemis Server. Please check your internet connection.",
             "reconnected": "You were reconnected to the Artemis Server."
         },
-        "outdatedAlert": "<b>Your Artemis version is outdated:</b> Please reload the browser page to prevent incompatibilities."
+        "outdatedAlert": "<b>Your Artemis version is outdated.</b> Please update to the latest version:",
```
Comment (krusche) [inline src/main/webapp/i18n/en/alert.json:8]:
```suggestion
        "outdatedAlert": "<b>Your Artemis version is outdated.</b> Please update to the latest version.",
```

File: src/main/webapp/app/core/interceptor/artemis-version.interceptor.ts [hunk 3]
```diff
@@ -34,4 +59,49 @@ export class ArtemisVersionInterceptor implements HttpInterceptor {
...
+     * We need to have this second option because the "checkForUpdate()" call sometimes starts to return false after a while, even though we didn't reload / update yet.
+     *
+     * @private
+     */
+    private checkForUpdates() {
+        // first update the service worker
+        this.updates
+            .checkForUpdate()
+            .then((updateAvailable: boolean) => {
+                if (this.hasSeenOutdatedInThisSession || updateAvailable) {
```
Comment (ole-ve) [inline src/main/webapp/app/core/interceptor/artemis-version.interceptor.ts:78]: I assume this is a typo
```suggestion
                if (this.hasBeenOutdatedInThisSession || updateAvailable) {
```

File: src/test/javascript/spec/interceptor/artemis-version.interceptor.spec.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        jest.useRealTimers();
+    });
+
+    beforeEach(() => {
+        swUpdate = {
+            isEnabled: true,
+            activated: EMPTY,
+            available: EMPTY,
+            checkForUpdate: () => of(true).toPromise(),
+            activateUpdate: () => of(true).toPromise(),
```
Comment (ole-ve) [inline src/test/javascript/spec/interceptor/artemis-version.interceptor.spec.ts:36]:
```suggestion
            checkForUpdate: () => Promise.resolve(true),
            activateUpdate: () => Promise.resolve(true),
```
I think this should work to resolve the client style issues.

**Review by ole-ve — APPROVED** [review]

File: src/test/javascript/spec/interceptor/artemis-version.interceptor.spec.ts (hunk 4 above)
Comment (bassner) [reply src/test/javascript/spec/interceptor/artemis-version.interceptor.spec.ts]: I've adressed this is in 71d127a :)

**Review by akesfeden — APPROVED** [review]
Tested on TS3, with prepared commit builds, alerts are shown persistently and Update button updates the version on the top left. Code changes and tests look good.

**Review by dveljacic — APPROVED** [review]
Manually tested on TS2. On Chrome browser, everything works exactly like it is described. Thanks for the detailed testing steps! ✅ 



However, on Firefox it is not working, at least didn't work for me. After the second deployment is started, the page shows an alert about being disconnected. After it's been successfully deployed, a green alert pops up that the connection has been reestablished, but the blue "your app is outdated" alert does not pop up. 

I assume it has to do something with the following error message in Firefox's console: `ERROR Error: Uncaught (in promise): Error: Service workers are disabled or not supported by this browser`. 

I don't know if this is a known issue that should be tackled in a different PR (if even possible) or is something related to my Firefox setup, but I thought it's worth mentioning.

**Comment by bassner** [comment]
@dveljacic Sounds like you have a browser setting or plugin that prevents service workers from being installed ... It works in my Firefox at least :)

**Review by willergeorg — APPROVED** [review]
Code changes lgtm.

**Review by b-fein — APPROVED** [review]
Code looks good. :+1:

**Pull request merged**